File: Lesson2/src/main/java/com/cnebrera/uc3/tech/lesson2/util/FixSizeMessage.java
package com.cnebrera.uc3.tech.lesson2.util;

import java.nio.ByteBuffer;
import java.util.Random;

/**
 * Represents a simple message
 */
public class FixSizeMessage
{
    /** Random number generator */
    private final static Random RND = new Random();

    /** Size of the message in binary, 8 bytes for double, 4 for each int */
    public final static int MSG_BINARY_SIZE = 16;

    /** Intrument ID */
    private final int instrument;

    /** Price */
    private final double price;

    /** Volume */
    private final int volume;

    public FixSizeMessage(int instrument, double price, int volume)
    {
        this.instrument = instrument;
        this.price = price;
        this.volume = volume;
    }

    /**
     * Convert the message into binary and store it in the given buffer.
     *
     * @param buffer the buffer that will store the serialized message
     */
    public void toBinary(final ByteBuffer buffer)
    {
        buffer.putInt(instrument);
        buffer.putDouble(price);
        buffer.putInt(volume);
    }

    /**
     * Read the msg from binary
     * @param buffer the buffer containing the message
     * @return the read message
     */
    public static FixSizeMessage readMsgFromBinary(final ByteBuffer buffer)
    {
        return new FixSizeMessage(
                buffer.getInt(),
                buffer.getDouble(),
                buffer.getInt());
    }

    /** Generates a random message */
    public static FixSizeMessage generateRandomMsg()
    {
        return new FixSizeMessage(RND.nextInt(), RND.nextDouble(), RND.nextInt());
    }

    @Override
    public String toString()
    {
        return "FixSizeMessage{" +
                "instrument=" + instrument +
                ", price=" + price +
                ", volume=" + volume +
                '}';
    }
}

File: Lesson2/src/models/FixSizeMessage.ts
import { randomInt } from 'crypto';

/**
 * Represents a fixed-size message
 * Each message contains an instrument (number), price (number), and volume (number).
 */
class FixSizeMessage {
    public static readonly MSG_BINARY_SIZE = 16;

    /**
     * Creates a new fixed-size message
     * @param instrument - Instrument ID as a number (integer BE, 4 bytes)
     * @param price - Price as a floating-point number (double BE, 8 bytes)
     * @param volume - Volume as a number (integer BE, 4 bytes)
     */
    constructor(public instrument: number, public price: number, public volume: number) { }

    /**
     * Serializes the message into binary format and writes it into the buffer
     * @param buffer - The buffer to store the binary data
     */
    public toBinary(buffer: Buffer): void {
        // TODO 1: Serialize the instrument, price, and volume into the buffer
        // Hint: Use `buffer.writeInt32BE` for integers and `buffer.writeDoubleBE` for the floating-point number.
        // Remember to increase the `offset` while you write to the `buffer`.
    }

    /**
     * Deserializes a message from binary format
     * @param buffer - The buffer containing the binary message (must be 16 bytes)
     * @returns A new FixSizeMessage instance
     */
    public static fromBinary(buffer: Buffer): FixSizeMessage {
        // TODO 2: Serialize the instrument, price, and volume into the buffer
        // Hint: Use `buffer.writeInt32BE` for integers and `buffer.writeDoubleBE` for the floating-point number.
        // Remember to increase the `offset` while you read to the `buffer`.
        // const instrument = ;
        // const price = ;
        // const volume = ;

        return new FixSizeMessage(instrument, price, volume);
    }

    /**
     * Generates a random fixed-size message
     * @returns A randomly generated FixSizeMessage instance
     */
    public static generateRandomMsg(): FixSizeMessage {
        return new FixSizeMessage(
            randomInt(1, 10000),
            parseFloat((Math.random() * 1000).toFixed(2)),
            randomInt(1, 10000),
        );
    }

    /**
     * Returns a string representation of the message
     * @returns A string describing the message
     */
    public toString(): string {
        return `FixSizeMessage(${FixSizeMessage.MSG_BINARY_SIZE}){instrument=${this.instrument}, price=${this.price}, volume=${this.volume}}`;
    }
}

export default FixSizeMessage;
